File: src/py_srvcli/py_srvcli/add_three_ints_server.py
from tutorial_interfaces.srv import AddThreeInts
import rclpy
from rclpy.node import Node

class MinimalService(Node):

    def __init__(self):
        super().__init__('add_three_ints_server')
        self.srv = self.create_service(AddThreeInts, 'add_three_ints', self.add_three_ints_callback)

    def add_three_ints_callback(self, request, response):
        response.sum = request.a + request.b + request.c
        self.get_logger().info(f'Incoming request: a={request.a}, b={request.b}, c={request.c}')
        return response

def main(args=None):
    rclpy.init(args=args)

    minimal_service = MinimalService()

    rclpy.spin(minimal_service)

    # Destroy the node explicitly
    minimal_service.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_srvcli/src/add_three_ints_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "tutorial_interfaces/srv/add_three_ints.hpp"

#include <memory>

void add(const std::shared_ptr<tutorial_interfaces::srv::AddThreeInts::Request> request,
          std::shared_ptr<tutorial_interfaces::srv::AddThreeInts::Response> response)
{
  response->sum = request->a + request->b + request->c;
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Incoming request: a=%ld, b=%ld, c=%ld", request->a, request->b, request->c);
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Sending response: sum=%ld", response->sum);
}

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("add_three_ints_server");
  auto service = node->create_service<tutorial_interfaces::srv::AddThreeInts>("add_three_ints", &add);
  
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Ready to add three ints.");
  
  rclcpp::spin(node);
  rclcpp::shutdown();
}